src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java

```diff
@@ -26,22 +49,81 @@ public KeyworkerStatsDto getStatsForStaff(Long staffId, String prisonId, LocalDa
 
-        Map<String, Integer> usageGroupedBySubType =  usageCounts.stream()
-                .collect(Collectors.groupingBy(CaseNoteUsageDto::getCaseNoteSubType,
-                        Collectors.summingInt(CaseNoteUsageDto::getNumCaseNotes)));
+        List<String> prisonerNosList = applicableAssignments.stream().map(OffenderKeyworker::getOffenderNo).distinct().collect(Collectors.toList());
 
-        Integer entryCount = usageGroupedBySubType.get("KA");
-        Integer sessionCount = usageGroupedBySubType.get("KS");
+        if (!prisonerNosList.isEmpty()) {
+            List<CaseNoteUsagePrisonersDto> usageCounts =
+                    nomisService.getCaseNoteUsageForPrisoners(prisonerNosList, KEYWORKER_CASENOTE_TYPE, null, fromDate, toDate);
 
+            Map<String, Integer> usageGroupedBySubType = usageCounts.stream()
+                    .collect(Collectors.groupingBy(CaseNoteUsagePrisonersDto::getCaseNoteSubType,
+                            Collectors.summingInt(CaseNoteUsagePrisonersDto::getNumCaseNotes)));
+
+            Integer sessionCount = usageGroupedBySubType.get(KEYWORKER_SESSION_SUB_TYPE);
+            int sessionsDone = sessionCount != null ? sessionCount : 0;
+
+            int projectedKeyworkerSessions = getProjectedKeyworkerSessions(applicableAssignments, staffId, prisonId, fromDate, nextEndDate);
+            final BigDecimal complianceRate = getComplianceRate(sessionsDone, projectedKeyworkerSessions);
+
+            Integer entryCount = usageGroupedBySubType.get(KEYWORKER_ENTRY_SUB_TYPE);
+
+            return KeyworkerStatsDto.builder()
+                    .projectedKeyworkerSessions(projectedKeyworkerSessions)
+                    .complianceRate(complianceRate)
+                    .caseNoteEntryCount(entryCount != null ? entryCount : 0)
+                    .caseNoteSessionCount(sessionsDone)
+                    .build();
+        }
         return KeyworkerStatsDto.builder()
                 .projectedKeyworkerSessions(0)
-                .complianceRate(0)
-                .caseNoteEntryCount(entryCount != null ? entryCount : 0)
-                .caseNoteSessionCount(sessionCount != null ? sessionCount :0)
+                .complianceRate(PERCENT_OPERAND)
+                .caseNoteEntryCount(0)
+                .caseNoteSessionCount(0)
                 .build();
     }
 
+    public PrisonStatsDto getPrisonStats(String prisonId, LocalDate fromDate, LocalDate toDate) {
+        Validate.notNull(prisonId,"prisonId");
+        Validate.notNull(fromDate, "fromDate");
+        Validate.notNull(toDate, "toDate");
+
+        return PrisonStatsDto.builder().build();
+    }
+
+    private BigDecimal getComplianceRate(int sessionCount, int projectedKeyworkerSessions) {
+        BigDecimal complianceRate = PERCENT_OPERAND;
+
+        if (projectedKeyworkerSessions > 0)  {
+            complianceRate = new BigDecimal(sessionCount).setScale(4, BigDecimal.ROUND_HALF_UP)
```

Comment:
This looks like a nice API for working calculations. Would using this approach benefit the logic in getProjectedKeyworkerSessions?